### ISSUE TITLE
UI::messagebox causes UI::start_timer to misfire

### ISSUE THREAD

**Issue by noelwarren** [SketchUp/api-issue-tracker#342]
SketchUp 2019 Windows

When a UI::messagebox is called from within a UI::start_timer block it causes the block to be called multiple times even though its second argument (repeat) is false.

The block should only be called once

Just copy and paste the following and wait a second.  Then try and play catch up with the windows that get created :)
```ruby
UI.start_timer(1, false) {
  puts "This should only fire once"
  UI.messagebox("Wait for more than a second to see the block get executed again")
}
```

It turns out we are aware of this and try to avoid calling message boxes from within these timers.  But unfortunately third party plugins call our extension from within timers and cause this issue.

**Comment by Nazz78** [comment]
I believe this is a case also with older versions of SketchUp.

**Comment by DanRathbun** [comment]
A workaround can be to use a `@count` variable ...
```ruby
@count = 0
UI.start_timer(1, false) {
  next if (@count += 1) > 1
  puts "This should only fire once"
  UI.messagebox("Wait for more than a second to see the block get executed again")
}
```
You can also stop the timer for good measure (but I think there was a recent bug with this?)
```ruby
@count = 0
@tid = UI.start_timer(1, false) {
  UI.stop_timer(@tid) if @count == 1
  next if (@count += 1) > 1
  puts "This should only fire once"
  UI.messagebox("Wait for more than a second to see the block get executed again")
}
```

**Comment by Nazz78** [comment]
@DanRathbun great tip, thanks for sharing!

**Comment by thomthom** [comment]
Yes, this has been known for a while. I think we looked at it a few years ago and it was for some reason non-trivial to fix. Probably worth looking into again because I have so many projects that use this pattern:

https://github.com/SketchUp/testup-2/blob/f83e12eb96d54691cdf2adfe917289702be517ae/src/testup/defer.rb#L11-L22

Short term we can at least document this bug in the docs with example workaround.

**Comment by noelwarren** [comment]
I didn't catch the note in the start_timer documentation about modal windows.  My bad.

As I mentioned we have 3rd party integrators calling our code (which uses messageboxes) from within blocks. All workarounds suggested use some kind of variable around the block scope.  This doesn't get propagated to our method calls so this isn't an option for us.

**Comment by thomthom** [comment]
> All workarounds suggested use some kind of variable around the block scope. This doesn't get propagated to our method calls so this isn't an option for us.

Why is it not an option to you? You mean the variation above with instance variables? Instance variables are not needed.

```rb
module TestUp
  module Execution

    def self.delay(seconds, &block)
      done = false
      UI.start_timer(seconds, false) do
        next if done
        done = true
        block.call
      end
    end

    def self.defer(&block)
      self.delay(0.0, &block)
    end

  end
end # module TestUp
```

The `done` variable is captured by the block and doesn't get affected by multiple calls to the `delay` method.

**Comment by noelwarren** [comment]
Hope this illustrates what I mean.  It is the fact we are getting called from another extension that makes this tricky to handle.
```module UnknownExtension
    def foo
      UI.start_timer(3, false) {
        OurExtension.bar
      }
  end
end

module OurExtension
  def bar
    UI.messagebox("Just a regular message (from the our perspective)")
  end
end
```
We have found a workaround for our particular use-case but it is very case specific

**Comment by thomthom** [comment]
Ah, I see. Yea, that is a challenge.

**Comment by DanRathbun** [comment]
Here's an idea ...
```ruby
module UnknownExtension
    def self.foo
      UI.start_timer(1, false) {
        OurExtension.bar
      }
  end
end

module OurExtension
  def self.bar
    from = caller_locations.first
    msg = "Just a regular message (from the perspective of OurExtension)"
    if from.label.start_with?('block') # <--<< Tests if call comes from a block
      # It's likley an External call
      UI::Notification.new(
        Object.new, # <-- Should be YOUR SketchupExtension object
        msg,
        Sketchup.find_support_file(
          'ShippedExtensions/su_dynamiccomponents/images/configurator_tool.png'
        ),
        "Whoops!"
      ).show
    else
      UI.messagebox(msg)
    end
  end
end
```

The `UI::Notification` is just an example of a non-modal element that notifies the user.

You can also use **non-modal** `UI::HtmlDialog` windows that **look and act** like messageboxes, instead of the modal native messagbox windows.

**Comment by kalpana-ghodeswar** [comment]
Hi [noelwarren](https://github.com/noelwarren), 

As part of trying to reproduce old bugs, I came across with this one. I tried this code with latest version SU2025.0 and it's working fine for me, it's triggering the message box only once.

`
UI.start_timer(1, false) {
  puts "This should only fire once"
  UI.messagebox("Wait for more than a second to see the block get executed again")
}
`

Is this still an issue for you? can you provide the code snippet please? 

Thanks

**Comment by noelwarren** [comment]
This is indeed fixed and no longer an issue

**Comment by kalpana-ghodeswar** [comment]
Thanks for confirming, closing the issue now :)
